engine/script/src/script_html5_js.cpp

```diff
@@ -97,9 +121,81 @@ namespace dmScript
         return 1;
     }
 
+    /*# set a JavaScript interaction listener callback from lua
+     *
+     * Set a JavaScript interaction listener callaback from lua that will be
+     * invoked when a user interacts with the web page by clicking, touching or typing.
+     * The callback can then call DOM restricted actions like requesting a pointer lock,
+     * or start playing sounds the first time the callback is invoked.
+     *
+     * @name html5.set_interaction_listener
+     * @param callback [type:function(self] The interaction callback. Pass an empty function or nil if you no longer wish to receive callbacks.
+     *
+     * `self`
+     * : [type:object] The calling script
+     *
+     * @examples
+     *
+     * ```lua
+     * local function on_interaction(self)
+     *     print("on_interaction called")
+     *     html5.set_interaction_listener(nil)
+     * end
+     *
+     * function init(self)
+     *     html5.set_interaction_listener(on_interaction)
+     * end
+     * ```
+     */
+    static int Html5_SetInteractionListener(lua_State* L)
+    {
+        luaL_checkany(L, 1);
+
+        if (lua_isnil(L, 1))
+        {
+            bool remove_js_callbacks = g_InteractionCallback != 0;
+            if (g_InteractionCallback)
+            {
+                EM_ASM({
+                    document.removeEventListener('click', Module.__defold_interaction_listener);
+                    document.removeEventListener('keyup', Module.__defold_interaction_listener);
+                    document.removeEventListener('touchend', Module.__defold_interaction_listener);
+                    Module.__defold_interaction_listener = undefined;
+                });
+                dmScript::DestroyCallback(g_InteractionCallback);
+            }
+            g_InteractionCallback = 0;
+
+            return 0;
+        }
+
+        if (g_InteractionCallback)
+        {
+            dmScript::DestroyCallback(g_InteractionCallback);
+        }
+
+        g_InteractionCallback = dmScript::CreateCallback(L, 1);
+        if (!dmScript::IsCallbackValid(g_InteractionCallback))
+        {
+            return luaL_error(L, "Failed to create callback");
+        }
+
+        EM_ASM({
+            Module.__defold_interaction_listener = function () {
+                _dmScript_RunInteractionCallback();
+            };
+            document.addEventListener('click',    Module.__defold_interaction_listener);
+            document.addEventListener('keyup',    Module.__defold_interaction_listener);
+            document.addEventListener('touchend', Module.__defold_interaction_listener);
```

Comment:
Q: The callbacks are set on the document which will respond to interacting anywhere on the page, is that what we want or do we want to restrict interaction to the canvas element only?

Reply:
I believe it is ok with interaction anywhere on the page...

Yes: https://developer.chrome.com/blog/autoplay/

"Chrome's autoplay policies are simple:

* Muted autoplay is always allowed.
* Autoplay with sound is allowed if:
    * The user has interacted with the domain (click, tap, etc.)."